seqr/views/utils/test_utils.py

```diff
@@ -91,6 +91,93 @@ def _check_login(self, url, permission_level, request_data=None):
 
         self.login_staff_user()
 
+
+ANVIL_WORKSPACES = [{
+    'workspace_namespace': 'test_namespace',
+    'workspace_name': 'anvil-1kg project n\u00e5me with uni\u00e7\u00f8de',
+    'public': False,
+    'acl': {
+        'test_user_manager@test.com': {
+            "accessLevel": "WRITER",
+            "pending": False,
+            "canShare": True,
+            "canCompute": True
+        },
+        'test_user_no_staff@test.com': {
+            "accessLevel": "READER",
+            "pending": False,
+            "canShare": True,
+            "canCompute": True
+        }
+    }
+}, {
+    'workspace_namespace': 'test_namespace',
+    'workspace_name': 'anvil-empty project',
+    'public': False,
+    'acl': {
+        'test_user_manager@test.com': {
+            "accessLevel": "WRITER",
+            "pending": False,
+            "canShare": True,
+            "canCompute": True
+        },
+        'test_user_no_staff@test.com': {
+            "accessLevel": "READER",
+            "pending": False,
+            "canShare": True,
+            "canCompute": True
+        }
+    }
+}, {
+    'workspace_namespace': 'test_namespace',
+    'workspace_name': 'anvil-project 1000 Genomes Demo',
+    'public': False,
+    'acl': {
+        'test_user_manager@test.com': {
+            "accessLevel": "WRITER",
+            "pending": False,
+            "canShare": True,
+            "canCompute": True
+        }
+    }
+}
+]
+
+
+def get_ws_acl_side_effect(url):
+    workspace_namespace, workspace_name = url.split('/')[2:4]
+    wss = filter(lambda x: x['workspace_namespace'] == workspace_namespace and x['workspace_name'] == workspace_name, ANVIL_WORKSPACES)
+    wss = list(wss)
+    return wss[0]['acl'] if wss else {}
+
+
+def get_workspaces_side_effect(user, fields):
+    return [
+        {
+            'public': ws['public'],
+            'accessLevel': ws['acl'][user.email]['accessLevel'],
+            'workspace':{
+                'namespace': ws['workspace_namespace'],
+                'name': ws['workspace_name']
+            }
+        } for ws in ANVIL_WORKSPACES if user.email in ws['acl'].keys()
+    ]
+
+
+class AnvilAuthenticationTestCase(AuthenticationTestCase):
+    @classmethod
+    def setUpTestData(cls):
+        cls.staff_user = User.objects.get(username='test_user')
+        cls.manager_user = User.objects.get(username='test_user_manager')
+        cls.collaborator_user = User.objects.get(username='test_user_non_staff')
+        cls.no_access_user = User.objects.get(username='test_user_no_access')
+
+        for project in Project.objects.all():
+            project.workspace_namespace = ANVIL_WORKSPACES[project.pk-1]['workspace_namespace']
+            project.workspace_name = ANVIL_WORKSPACES[project.pk - 1]['workspace_name']
+            project.save()
```

Comment:
I STRONGLY dislike this. It's incredibly confusing to load a fixture and then modify it, especially to modify it in an inherited class so it's not even clear that that is happening. I think you need to set up the fixture data to have the correct data. I would rather have the anvil and local test cases differ from each other a bit than do a hack like this

An easy way to allow the tests to be reused with the different login mechanisms but have slightly different results is to add class level constants, so instead if `self.assertEqual(len(response_json['projectsByGuid']), 2)` do `self.assertEqual(len(response_json['projectsByGuid']), NUM_COLLABORATOR_PROJECTS)` and then each of you class definitions define a different value for `NUM_COLLABORATOR_PROJECTS`

Reply:
Good idea to add login-mechanism-dependent results to the authentication class. I was hesitating about updating the fixture data. I could add the workspace information into the fixture data directly. But, I have the concern that it may not cover the cases of no workspace configurations for the local and mix tests. I really don't like to duplicate the fixture data for different tests. If I can use different predictions then it solves the problem.

Reply:
I think add workspaces to project fixture 1 but not to project 2 or 3, or maybe to projects 1 and 2 but not 3 or 1 and 3 but not 2, whichever makes for the best tests. That will allow you to test all 3 different cases without needing to modify the fixtures in the tests.

Reply:
I added workspaces to project 1 and 3, not 2.